src/coap_gnutls.c

```diff
@@ -1409,18 +1420,40 @@ int coap_dtls_send(coap_session_t *c_session,
 }
 
 int coap_dtls_is_context_timeout(void) {
-  return 1;
+  return 0;
 }
 
 coap_tick_t coap_dtls_get_context_timeout(void *dtls_context UNUSED) {
   return 0;
 }
 
-coap_tick_t coap_dtls_get_timeout(coap_session_t *c_session UNUSED) {
+coap_tick_t coap_dtls_get_timeout(coap_session_t *c_session, coap_tick_t now) {
+  coap_gnutls_env_t *g_env = (coap_gnutls_env_t *)c_session->tls;
+
+  if (g_env && g_env->g_session) {
+    unsigned int rem_ms = gnutls_dtls_get_timeout(g_env->g_session);
+
+    return now + rem_ms;
+  }
+
   return 0;
 }
 
-void coap_dtls_handle_timeout(coap_session_t *c_session UNUSED) {
+void coap_dtls_handle_timeout(coap_session_t *c_session) {
+  coap_gnutls_env_t *g_env = (coap_gnutls_env_t *)c_session->tls;
+
+  assert(g_env != NULL);
+  g_env->doing_dtls_timeout = 1;
+  if (((c_session->state == COAP_SESSION_STATE_HANDSHAKE) &&
+       (++c_session->dtls_timeout_count > c_session->max_retransmit)) ||
+      (do_gnutls_handshake(c_session, g_env) < 0)) {
```

Comment:
I am wondering if this should be `<= 0`. In case where`do_gnutls_handshake()` returns 0, and `receive_timeout()` would not set the writefds and exceptfds.

Reply:
Simple answer is no.  If the change is made, when 0 is returned (handshake incomplete), then coap_session_disconnected(c_session, COAP_NACK_TLS_FAILED) is called.  However, if there is an error (-1 returned), then GnuTLS has died and we need to close down the session.

We need to to a gnutls_handshake() to trigger the resend of the Client Hello, and receive_timeout() has to return "timed-out" for this to happen - hence no writefds etc. being set.

However, in normal setup packet interchange, GnuTLS uses receive_timeout() (set up by gnutls_transport_set_pull_timeout_function()) to see if it is able to "write" out the next exchange message - function naming is not ideal here!

Reply:
Ok, thank you for the clarification!